vhdl_syntax/src/parser/productions/names.rs

```diff
@@ -32,4 +74,480 @@ impl<T: TokenStream> Parser<T> {
+                self.end_node();
+            } else {
+                // TODO: subtype_indication or expression?
+                self.start_node(Internal(SubtypeIndicationOrExpressionTokens));
+                self.expect_token(LeftPar);
+                match self.distance_to_closing_paren() {
+                    Some(distance) => self.skip_n(distance),
+                    None => {
+                        self.end_node();
+                        self.eof_err();
+                        return;
+                    }
+                }
+                self.expect_token(RightPar);
+                self.end_node();
+            }
+
+            self.name_tail();
+        } else if is_start_of_attribute_name(self) {
+            self.start_node(AttributeName);
+            if self.next_is(LeftSquare) {
+                self.signature();
+            }
+            self.expect_token(Tick);
+
+            // `range` is a keyword, but may appear as a `attribute_name`
+            if !self.opt_identifier() {
+                self.expect_kw(Kw::Range);
+            }
+
+            if self.next_is(LeftPar) {
+                self.start_node(ParenthesizedExpression);
+                self.expect_token(LeftPar);
+                self.expression();
+                self.expect_token(RightPar);
+                self.end_node();
+            }
+            self.end_node();
+            self.name_tail();
+        }
+    }
+
+    pub fn external_name(&mut self) {
+        // LRM §8.7
+        self.start_node(ExternalName);
+        self.expect_token(LtLt);
+
+        self.expect_one_of_tokens([
+            Keyword(Kw::Constant),
+            Keyword(Kw::Signal),
+            Keyword(Kw::Variable),
+        ]);
+        self.external_pathname();
+        self.expect_token(Colon);
+        self.subtype_indication();
+
+        self.expect_token(GtGt);
+        self.end_node();
+    }
+
+    fn external_pathname(&mut self) {
+        // LRM §8.7
+        self.start_node(ExternalPathName);
+        match self.peek_token() {
+            Some(CommAt) => {
+                self.expect_token(CommAt);
+                self.identifier();
+                self.expect_token(Dot);
+                self.identifier();
+                self.expect_token(Dot);
+                self.identifier();
+                while self.opt_token(Dot) {
+                    self.identifier();
+                }
+            }
+            Some(Dot) => {
+                self.expect_token(Dot);
+                self.partial_pathname();
+            }
+            Some(Circ | Identifier) => {
+                while self.opt_token(Circ) {
+                    self.expect_token(Dot);
+                }
+                self.partial_pathname();
+            }
+            Some(_) => {
+                self.expect_tokens_err([CommAt, Dot, Circ, Identifier]);
+            }
+            None => {
+                self.eof_err();
+            }
+        }
```

Comment:
This (and probably some others) could be simplified by using the `match_next_token` resp. `match_next_token_consume` macro.